File: iosApp/iosApp/Screens/Login/LoginView.swift
//
//  LoginView.swift
//  iosApp
//
//  Created by Ashkeyev Ali on 16.07.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI
import SharedSDK

struct LoginView: View {
            
    let viewState: LoginViewState
    let eventHandler: (LoginEvent) -> Void

    var body: some View {
        VStack {
            VStack {
                Spacer().frame(height: 36)
                Text("Login Now")
                    .foregroundColor(.textPrimary)
                    .fontWeight(.bold)
                    .font(.system(size: 24))
                Text("Welcome back to PlayZone! Enter your email addres and your password to enjoy the latest features of PlayZone")
                    .foregroundColor(.textPrimary.opacity(0.5))
                    .fixedSize(horizontal: false, vertical: true)
                    .multilineTextAlignment(.center)
                    .padding(EdgeInsets(top: 16, leading: 30, bottom: 0, trailing: 30))
                
                Spacer().frame(height: 50)
                
                CommonTextField(hint: "Your login", enabled: !viewState.isSending, isSecure: false) {newValue in
                    eventHandler(.EmailChanged(email: newValue))
                    }
                Spacer().frame(height: 24)
                
                CommonTextField(hint: "Password", enabled: !viewState.isSending, isSecure: viewState.passwordHidden) {newValue in
                    eventHandler(.PasswordChanged(password: newValue))
                }
            }
            
            LoginActionView(
                viewState: viewState,
                onForgotClicked: {
                    eventHandler(.ForgotPasswordClicked())
                },
                
                onSubmitClicked: {
                    eventHandler(.LoginClicked())
                }).frame(height: 56)
            
            Spacer()
            HStack {
                Text("Don't have account?")
                    .foregroundColor(.textPrimary)
                    .opacity(0.5)
                Text("Create one")
                    .foregroundColor(.tint)
                    .fontWeight(/*@START_MENU_TOKEN@*/.bold/*@END_MENU_TOKEN@*/)
                    .onTapGesture {
                        eventHandler(.RegisterClicked())
                        
                    }
            }
        }

    }
}

struct LoginActionView: View {
    let viewState: LoginViewState
    let onForgotClicked: () -> Void
    let onSubmitClicked: () -> Void
    var body: some View {
        VStack {
            Spacer().frame(height: 30)
            HStack {
                Spacer()
                Text("Forgot password")
                    .foregroundColor(.tint)
                    .onTapGesture{onForgotClicked()}
                Spacer().frame(width: 30)

            }
            Spacer().frame(height: 30)
            
            ActionButton(title: "Login now", enabled: !viewState.isSending){
                onSubmitClicked()
            }


            
        }
        
    }
}

struct LoginView_Previews: PreviewProvider {
    static var previews: some View {
        LoginView( viewState: LoginViewState(email: "ashkeyevli", password: "password", isSending: false, passwordHidden: true), eventHandler: { event in }
        ).background(Color.backgroundPrimary)
    }
}

File: common/auth/compose/src/androidMain/kotlin/login/LoginView.kt
package login

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Button
import androidx.compose.material.ButtonDefaults
import androidx.compose.material.Icon
import androidx.compose.material.Text
import androidx.compose.material.TextField
import androidx.compose.material.TextFieldDefaults
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Clear
import androidx.compose.material.icons.filled.Lock
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import login.models.LoginEvent
import login.models.LoginViewState
import theme.Theme

@Composable
fun LoginView(state: LoginViewState, eventHandler: (LoginEvent) -> Unit) {
    Column(modifier = Modifier.padding(30.dp), horizontalAlignment = Alignment.CenterHorizontally){
        Text("Login Now", color = Theme.colors.secondaryTextColor,
            fontSize = 24.sp, fontWeight = FontWeight.Bold, )
        Text(
            modifier = Modifier.padding(top = 15.dp),
            text = "Welcome back to PlayZone! Enter your email addres and your password to enjoy the latest features of PlayZone",
            color = Theme.colors.hintTextColor, fontSize = 14.sp, textAlign = TextAlign.Center)

        TextField(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp),
            value = state.email,
            enabled = !state.isSending,
            colors = TextFieldDefaults.textFieldColors(
                backgroundColor = Color(0xFF1F2430),
                textColor = Color(0xFF696C75),
                cursorColor = Theme.colors.highlightTextColor,
                focusedIndicatorColor = Color.Transparent,
                unfocusedIndicatorColor = Color.Transparent
            ),
            placeholder = { Text(text = "Your login", color = Theme.colors.hintTextColor) },
            shape = RoundedCornerShape(10.dp),
            onValueChange = {
                eventHandler(LoginEvent.EmailChanged(it))
            })

        Spacer(modifier = Modifier.height(24.dp))

        TextField(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp),
            value = state.password,
            enabled = !state.isSending,
            colors = TextFieldDefaults.textFieldColors(
                backgroundColor = Color(0xFF1F2430),
                textColor = Color(0xFF696C75),
                cursorColor = Theme.colors.highlightTextColor,
                focusedIndicatorColor = Color.Transparent,
                unfocusedIndicatorColor = Color.Transparent
            ),
            visualTransformation = if (state.passwordHidden) PasswordVisualTransformation() else VisualTransformation.None,
            placeholder = { Text(text = "Your password", color = Theme.colors.hintTextColor) },
            trailingIcon = {
                Icon(
                    imageVector = if (state.passwordHidden) Icons.Default.Lock else Icons.Default.Clear,
                    contentDescription = "Password hidden",
                    modifier = Modifier.clickable {
                        eventHandler(LoginEvent.PasswordShowClicked)
                                                  },
                    tint = Theme.colors.hintTextColor,
                )
            },
            shape = RoundedCornerShape(10.dp),
            onValueChange = {
                eventHandler(LoginEvent.PasswordChanged(it))
            })

        Spacer(modifier = Modifier.height(84.dp))

        Row(modifier = Modifier.fillMaxWidth()) {
            Spacer(modifier = Modifier.weight(1f))
            Text(
                text = "Forgot your password?",
                color = Theme.colors.primaryAction,
                fontSize = 12.sp,
                modifier = Modifier.clickable { eventHandler(LoginEvent.ForgotPasswordClicked) }
            )
        }
        Spacer(modifier = Modifier.height(30.dp))




        Button(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp),
            colors = ButtonDefaults.buttonColors(
                backgroundColor = Theme.colors.primaryAction
            ),
            enabled = !state.isSending,
            shape = RoundedCornerShape(10.dp),
            onClick = { eventHandler(LoginEvent.LoginClicked) }) {
            Text(text = "Login Now", color = Theme.colors.primaryTextColor, fontSize = 16.sp, fontWeight = FontWeight.Bold)

        }

    }


}